File: hacking_with_perl/web_chache_deception.pl
#!/usr/bin/perl

use strict;
use warnings;
use LWP::UserAgent;
use HTTP::Request;
use HTTP::Headers;
use HTTP::Cookies;

# Define ANSI escape sequences for color formatting
my $color_red = "\e[31m";
my $color_green = "\e[32m";
my $color_reset = "\e[0m";

# Prompt for target website URL
print "Enter your target website URL: ";
my $url = <STDIN>;
chomp($url);

# Create user agent object with SSL certificate verification
my $ua = LWP::UserAgent->new(
    ssl_opts => {
        verify_hostname => 1,
    }
);
$ua->timeout(10);  # Set timeout to 10 seconds
$ua->agent("Mozilla/5.0");  # Set a custom User-Agent string

# Enable cookie handling
$ua->cookie_jar(HTTP::Cookies->new);

# Create HTTP request object
my $request = HTTP::Request->new(GET => $url);
$request->header('Accept' => 'text/html');  # Specify accepted content type

# Send request to target URL
my $response = $ua->request($request);

# Check for redirections
if ($response->is_redirect) {
    my $redirect_url = $response->header('Location');
    print "Redirected to: $redirect_url\n";
    $request = HTTP::Request->new(GET => $redirect_url);
    $response = $ua->request($request);
}

# Check for errors
if ($response->is_success) {
    # Check for Web Cache Deception vulnerability
    if ($response->header('X-Cache')) {
        print $color_red, "The website is vulnerable to Web Cache Deception.\n", $color_reset;
        print "X-Cache Header: ", $response->header('X-Cache'), "\n";
    } else {
        print $color_green, "The website is not vulnerable to Web Cache Deception.\n", $color_reset;
    }
    # Print response details
    print "HTTP Response Code: ", $response->code, "\n";
    print "Response Headers:\n", $response->headers_as_string, "\n";
} else {
    print $color_red, "Error: ", $response->status_line, "\n", $color_reset;
}

File: hacking_with_php/web_chache_deception.php
<?php

// Define ANSI color escape codes
const COLOR_RED = "\033[0;31m";
const COLOR_GREEN = "\033[0;32m";
const COLOR_RESET = "\033[0m";

// Function to send HTTP request and return response headers
function sendHttpRequest($url, $method = 'GET', $headers = array()) {
    $options = array(
        'http' => array(
            'method' => $method,
            'header' => implode("\r\n", $headers),
        )
    );
    $context = stream_context_create($options);
    return get_headers($url, 1, $context);
}

// Function to check for Web Cache Deception
function checkWebCacheDeception($url, $method, $modification, $followRedirects, $useCookies, $userAgent) {
    // Send a request with the modified URL
    $modifiedUrl = $url . $modification;
    $headers = array();

    // Add cookies if requested
    if ($useCookies) {
        $headers[] = 'Cookie: test=1'; // Replace 'test=1' with actual cookies if needed
    }

    // Add User-Agent header
    $headers[] = 'User-Agent: ' . $userAgent;

    // Follow redirects if requested
    if ($followRedirects) {
        $response = followRedirections($modifiedUrl, $method, $headers);
    } else {
        $response = sendHttpRequest($modifiedUrl, $method, $headers);
    }

    // Check if the response contains the original URL
    $originalUrl = isset($response['Location']) ? $response['Location'] : null;
    if ($originalUrl === $url) {
        echo COLOR_RED . "Vulnerable to Web Cache Deception: Original URL detected in response.\n" . COLOR_RESET;
        echo "Detected Location Header: $originalUrl\n";
    } else {
        echo COLOR_GREEN . "Not vulnerable to Web Cache Deception.\n" . COLOR_RESET;
    }
}

// Function to follow redirections recursively and return final response headers
function followRedirections($url, $method = 'GET', $headers = array(), $maxRedirects = 5) {
    $response = sendHttpRequest($url, $method, $headers);

    // Check if redirection occurred
    if (isset($response['Location']) && $maxRedirects > 0) {
        return followRedirections($response['Location'], $method, $headers, $maxRedirects - 1);
    }

    return $response;
}

// Prompt user for the target website URL
echo "Enter your target website URL: ";
$targetUrl = trim(fgets(STDIN));

// Prompt user for the HTTP method
echo "Enter HTTP method (GET/POST): ";
$method = strtoupper(trim(fgets(STDIN)));

// Prompt user for the URL modification string
echo "Enter URL modification string (e.g., /index.php%3fa%3d1): ";
$modification = trim(fgets(STDIN));

// Prompt user whether to follow redirects
echo "Follow redirects? (yes/no): ";
$followRedirectsInput = strtolower(trim(fgets(STDIN)));
$followRedirects = ($followRedirectsInput === 'yes');

// Prompt user whether to include cookies
echo "Include cookies? (yes/no): ";
$useCookiesInput = strtolower(trim(fgets(STDIN)));
$useCookies = ($useCookiesInput === 'yes');

// Prompt user for the User-Agent header
echo "Enter User-Agent header (press Enter to use default): ";
$userAgent = trim(fgets(STDIN));

// Use default User-Agent if not provided
if (empty($userAgent)) {
    $userAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36";
}

// Test the target website for Web Cache Deception
echo "\nTesting $targetUrl for Web Cache Deception...\n";
checkWebCacheDeception($targetUrl, $method, $modification, $followRedirects, $useCookies, $userAgent);

?>
